### PR TITLE
Introduce new protocol-util module

### LINKED ISSUE #8837: Protocol Jackson improvements
**Description**

There are some existing issues with the current implementation of the `protocol-jackson` module. The main issues are:

1. For every type created in `protocol`, we need to add a corresponding type in `protocol-jackson`.
  1. This is error prone (as example, we already have missing types), and it's not so easy to add tests to guarantee this (though do-able)
  1. This adds unnecessary overhead for developers
1. The `Immutable` implementations are hidden - this was to reduce the API surface, but it turns out that this blocks, to some extent, interoperability with other tools and libraries, especially those operating by reflection (e.g. fixture/data generators)
1. Each of the types added in `protocol-jackson` need to also define how to deserialize nested types. This is also error prone and adds overhead for developers, and difficult to ensure via tests.
1. Copying into an immutable structure is a shallow copy, as Immutables doesn't recognize that, say, `JobRecordValue` is equivalent to `AbstractJobRecordValue`.

The proposed solution is:

- Annotate the existing types in the `protocol` module directly. These annotations are compile-time only, and as such, there is no extra dependency added to the module. 
- We combine all required annotations in a single one: `ZeebeProtocolValue` (or even just `ProtocolValue` - or suggest anything). This not only makes it easier to guarantee all types are correctly annotated, but also opens up the way for #8886 and other properties we would like to enforce in our protocol
- We would not add the Jackson bindings there, however, but instead keep `protocol-jackson` specifically to define a `ZeebeProtocolModule`, as is common with Jackson to define custom serialization.
- A new module will be created, `protocol-util`, which would add utilities to help using the protocol, notably make explicit the mapping of `ValueType`, `Intent`, and `T extends RecordValue`, and also keep track of the mapping between the interface types (e.g. `JobRecordValue`), their immutable counter part (e.g. `ImmutableJobRecordValue`), and their builders (e.g. `ImmutableJobRecordValue.Builder`). This is then used by the `protocol-jackson` module and any other module which wants this information (e.g. `protocol-test-util`).

This has several advantages:

1. Copying is now a deep copy, as you would expect
1. You can more easily use the `Immutable*` variants when you need to, and if you don't care about the serialization part, you can still use them
1. We can easily add a `ArchUnit` test which checks that the interface types are correctly annotated (which guarantees they will generate an immutable equivalent)
1. The mapping of `ValueType`, `Intent`, and values is made explicit - this can be further improved via #8886
1. It makes it very easy to add a random record generator once we have `protocol-util` and the immutable variants
1. The `protocol-jackson` module now is better defined - it's literally just a single Jackson module which you add to your object mapper and now knows how to deserialize the Zeebe protocol.
1. Having a random record factory which you can semi control (i.e. reusing `protocol-util`) makes it much easier to write exporter unit tests, where the data in the records isn't so important in most cases

See the draft PR for an idea of what it would look like.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

Adds a new module for protocol utilities, which for now provides a centralized way to discover type mappings, e.g. mapping value type to values or intents, or abstract protocol types to concrete protocol types.

These will be used in `protocol-jackson` in the next step, and later on in other modules, e.g. to generate random records.

## Related issues

related to #8837 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

File: protocol-util/pom.xml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    <groupId>io.camunda</groupId>
+    <artifactId>zeebe-parent</artifactId>
+    <version>1.4.0-SNAPSHOT</version>
+    <relativePath>../parent/pom.xml</relativePath>
+  </parent>
+
+  <artifactId>zeebe-protocol-util</artifactId>
+  <packaging>jar</packaging>
+
+  <name>Zeebe Protocol Utilities</name>
```
Comment (korthout) [inline protocol-util/pom.xml:16]: I'm missing some documentation on what this module is intended for (I know I know, it's in the commit message). I think this is the type of info where we can go for more prominent documentation, like:

🔧 Please provide a description tag
🔧 Please add a README.md (can be short)

**Review by korthout — APPROVED** [review]
LGTM 👍 I don't really have new comments, as I'd already reviewed this once before.

Please have a look at my suggestions.

File: protocol-util/pom.xml (hunk 1 above)
Comment (npepinpe) [reply protocol-util/pom.xml:16]: Do you mean like the `protocol`'s README?

I quote:

> # Zeebe Protocol
>
> Binary protocol of Zeebe.

This is almost poetry :smile:

Comment (korthout) [reply protocol-util/pom.xml:16]: So it would be something like this? 🤣 

> # Zeebe Protocol Util
> 
> Utilities for the binary protocol of Zeebe

Comment (npepinpe) [reply protocol-util/pom.xml:16]: Added one

**Comment by npepinpe** [comment]
bors merge

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**